### ISSUE TITLE
Replace usage of t.p.compat.reduce in twisted.names

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6563]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#6563|
|Type|task|
|Created|2013-06-10 18:34:12Z|
|Branch|https://github.com/twisted/twisted/tree/reduce-names-6563|

The builtin `reduce()` was moved to [functools.reduce](http://docs.python.org/2/library/functools.html#functools.reduce) in python2.6 and up and the builtin was removed in Python 3.

We can replace the usage of `twisted.python.compat.reduce` (introduced in r27667 for compatibility with python < 2.6) and usage of the builtin `reduce` with `functools.reduce` instead and eventually deprecate it.

```
twisted/names/srvconnect.py:13:from twisted.python.compat import reduce, unicode
twisted/names/srvconnect.py:161:        weightSum = reduce(lambda x, y: (None, x[1]+y[1]), weightIndex, (None, 0))[1]
twisted/names/test/test_names.py:23:from twisted.python.compat import reduce
twisted/names/test/test_names.py:404:        results = [copy.copy(r) for r in reduce(operator.add, test_domain_com.records.values())]
```

<details><summary>Searchable metadata</summary>

```
trac-id__6563 6563
type__task task
reporter__thijs thijs
priority__low low
milestone__Python_3_x Python-3.x
branch__branches_reduce_names_6563 branches/reduce-names-6563
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__names names
keywords__easy easy
time__1370889252000000 1370889252000000
changetime__1465472567761217 1465472567761217
version__None None
owner__thijs thijs
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

(In [40142]) Branching to 'reduce-names-6563'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

(In [40143]) replace t.p.compat.reduce calls in names, refs [#6563](https://github.com/twisted/twisted/issues/6563)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/reduce-names-6563).

Also fixed some pyflakes warnings; removed a `rand` var that was not being used.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj set owner to @thijstriemstra|
|-|-|

Thanks Thijs,

Notes:
 * Merges cleanly
 * Build results look good
 * Tests all pass locally
 * The changed lines seem to have coverage and the lines that use reduce also
   have coverage.

I had difficulty understanding the use of reduce in srvconnect and wonder if it
would be better replaced with a for loop, but that's for another ticket.

I think the reduce in test_names is also confusing, it's combining multiple
lists, and I think it could be replaced with sum, but again that's probably
something for another ticket.

```
In [5]: sum([[1], [2], [3]], [])
Out[5]: [1, 2, 3]
```

That unused rand variable looks suspicious; I wonder if it is meant to be used
in the SRV host selection. But I checked back to the original in r4762 and it
seems never to have been used.

Anyway, the branch looks good and it's a nice small diff.

So please merge.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [40144]) Merge reduce-names-6563: Replace usage of twisted.python.compat.reduce in twisted.names in favor of functools.reduce.

Author: thijs
Reviewer: rwall
Fixes: [#6563](https://github.com/twisted/twisted/issues/6563)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_6"></a>|@ralphm commented|
|-|-|

In changeset 6000696d16c2f3b5f94c240e82a85eb234352dcd
```
#!CommitTicketReference repository="" revision="6000696d16c2f3b5f94c240e82a85eb234352dcd"
Reimplement `pickServer` to conform to RFC 2782.

This brings back the call to `random.randint` that was removed as part
of an earlier change (see #6563), because it wasn't actually used.
```
